### PR TITLE
8329760: Add indexOf(Predicate<? super E> filter) to java.util.List interface

### PR DESCRIPTION
**Pull request by Evemose**
**Subject**
Addition of Predicate-based `indexOf` and `lastIndexOf` methods to `java.util.List`

**Motivation**
The motivation behind this proposal is to enhance the functionality of the `List` interface by providing a more flexible way to find the index of an element. Currently, the `indexOf` and `lastIndexOf` methods only accept an object as a parameter. This limits the flexibility of these methods as they can only find the index of exact object matches.

The proposed methods would accept a `Predicate` as a parameter, allowing users to define a condition that the desired element must meet. This would provide a more flexible and powerful way to find the index of an element in a list.

Here is a brief overview of the changes made in this pull request:

1. Added the `indexOf(Predicate<? super E> filter)` method to the `List` interface.
2. Added the `lastIndexOf(Predicate<? super E> filter)` method to the `List` interface.
3. Implemented these methods in all non-abstract classes that implement the `List` interface.

The changes have been thoroughly tested to ensure they work as expected and do not introduce any regressions. The test cases cover a variety of scenarios to ensure the robustness of the implementation.

For example, consider the following test case:
```
List<String> list = new ArrayList<>();
list.add("Object one");
list.add("NotObject two");
list.add("NotObject three");

int index1 = list.indexOf(s -> s.contains("ct t"));
System.out.println(index1); // Expected output: 1
int index2 = list.lastIndexOf(s -> s.startsWith("NotObject"));
System.out.println(index2); // Expected output: 2
```

Currently, to achieve the same result, we would have to use a more verbose approach:
```
int index1 = IntStream.range(0, list.size())
                     .filter(i -> list.get(i).contains("ct t"))
                     .findFirst()
                     .orElse(-1);
System.out.println(index1); // Output: 1
int index2 = IntStream.range(0, list.size())
                         .filter(i -> list.get(i).startsWith("NotObject"))
                         .reduce((first, second) -> second)
                         .orElse(-1);
System.out.println(index2); // Output: 2
```

I believe these additions would greatly enhance the functionality and flexibility of the `List` interface, making it more powerful and user-friendly. I look forward to your feedback and am open to making any necessary changes based on your suggestions.

Thank you for considering this proposal.

Best regards

PS: In proccess, indexOf(Predicate) was renamed to findIndex(Predicate) to preserve source compatibility

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change requires a CSR request matching fixVersion 24 to be approved (needs to be created)

### Issue
 * [JDK-8329760](https://bugs.openjdk.org/browse/JDK-8329760): Add indexOf(Predicate&lt;? super E&gt; filter) to java.util.List interface (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18639/head:pull/18639` \
`$ git checkout pull/18639`

Update a local copy of the PR: \
`$ git checkout pull/18639` \
`$ git pull https://git.openjdk.org/jdk.git pull/18639/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18639`

View PR using the GUI difftool: \
`$ git pr show -t 18639`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18639.diff">https://git.openjdk.org/jdk/pull/18639.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18639#issuecomment-2065397050)

### CODE REVIEW

**Comment by Evemose** [comment]
currently awaiting for initial issue triage to complete

**Comment by Evemose** [comment]
side note: looks like this code have been reformated and some unused imports has been removed. jdk compiles and works just fine, so i guess its not a big deal

**Comment by liach** [comment]
Hello @Evemose, if we add an API we should also consider if implementation can efficiently implement it. How is this new API better than using a ListIterator directly so it's worth the additional methods?

**Comment by Evemose** [comment]
> Hello @Evemose, if we add an API we should also consider if implementation can efficiently implement it. How is this new API better than using a ListIterator directly so it's worth the additional methods?

Hi! The implementations i provided in this pull requests are pretty much imitates behaviour of indexOf(Object). Every class that has overriden indexOf(Object) recieved implementation of indexOf(Predicate) that works the same way except for search is based on predicate and not equality. Therefore, the effectiveness of implementation is straight up depends on effectiveness of indexOf(Object) for each separatse java.util.List impl.

As for ListIterator, where suitable, it has already been used in implementations where it was considered suitable for indexOf(Object). The main goal of this pull requst is to provide shorthand for developers as search for index of element that match predicate is very common task and providing way to do it as comfortable as possible should vastly enhance development expirience.

Summarizing, its not that in current stdlib there are no ways to search for index of element based on condition, but this addition would be a great way to make Java development more comfortable and fast.

File: src/java.base/share/classes/java/util/AbstractList.java
Comment (ExE-Boss) [inline src/java.base/share/classes/java/util/AbstractList.java:221]:
```suggestion
```

File: src/java.base/share/classes/java/util/AbstractList.java
Comment (ExE-Boss) [inline src/java.base/share/classes/java/util/AbstractList.java:269]:
```suggestion
```

File: src/java.base/share/classes/java/util/List.java
Comment (ExE-Boss) [inline src/java.base/share/classes/java/util/List.java:702]:
```suggestion
    /**
     * Returns the index of the first occurrence of matching element
     * in this list, or -1 if this list does not contain the element.
     * More formally, returns the lowest index {@code i} such that
     * {@code filter.test(get(i))},
     * or -1 if there is no such index.
     *
     * @implSpec
     * This implementation first gets a list iterator (with
     * {@code listIterator()}).  Then, it iterates over the list until a
     * matching element is found or the beginning of the list is reached.
     *
     * @param filter a predicate to search for
     * @return the index of the first occurrence of the specified element in
     *         this list, or -1 if this list does not contain the element
     * @throws NullPointerException if passed filter is null
     */
    default int indexOf(Predicate<? super E> filter) {
        Objects.requireNonNull(filter);
        ListIterator<E> it = listIterator();
        while (it.hasNext()) {
            E e = it.next();
            if (filter.test(e)) {
                return it.previousIndex();
            }
        }
        return -1;
    }

    /**
     * Returns the index of the last occurrence of matching element
     * in this list, or -1 if this list does not contain the element.
     * More formally, returns the highest index {@code i} such that
     * {@code Objects.equals(o, get(i))},
     * or -1 if there is no such index.
     *
     * @implSpec
     * This implementation first gets a list iterator that points to the end
     * of the list (with {@code listIterator(size())}).  Then, it iterates
     * backwards over the list until the matching element is found, or the
     * beginning of the list is reached.
     *
     * @param filter a predicate to search for
     * @return the index of the last occurrence of the specified element in
     *         this list, or -1 if this list does not contain the element
     * @throws NullPointerException if passed filter is null
     */
    default int lastIndexOf(Predicate<? super E> filter) {
        Objects.requireNonNull(filter);
        ListIterator<E> it = listIterator(size());
        while (it.hasPrevious()) {
            E e = it.previous();
            if (filter.test(e)) {
                return it.nextIndex();
            }
        }
        return -1;
    }
```

**Review by ExE-Boss — CHANGES_REQUESTED** [review]
Newly added methods to existing widely‑implemented non‑sealed and non‑preview interfaces in `java.base` must be `default`, so move the implementations from `AbstractList` to `List`:

**Comment by Evemose** [comment]
@ExE-Boss Commited changes. Also, i see you added not-null assertion at the beginning of impl. If this is a desired behaviour, I could also add this to other impls

File: src/java.base/share/classes/java/util/AbstractList.java
Comment (ExE-Boss) [inline src/java.base/share/classes/java/util/AbstractList.java:29]: This is now unused:
```suggestion
```

**Comment by Evemose** [comment]
@ExE-Boss removed unused import

**Comment by xxDark** [comment]
Won't this make some calls ambiguous?
For example, passing `null`:
```Java
List<String> l;
l.indexOf(""); // Fine
l.indexOf(null); // Now fails to compile
```

**Comment by Evemose** [comment]
> Won't this make some calls ambiguous? For example, passing `null`:
> 
> ```java
> List<String> l;
> l.indexOf(""); // Fine
> l.indexOf(null); // Now fails to compile
> ```
> 
> Edit: I'm not a reviewer, but I decided to point this out.

I`ve tested it. Now calling indexOf with null as literal calls indexOf(Predicate) instead, as Predicate is lower in class hierarchy that Object (ofc it does not apply to varibles that store null, this works just as usual)

**Comment by liach** [comment]
A few remarks:
1. The concern on `indexOf(Object)` and `indexOf(Predicate)` is valid; `Optional` has `orElse` vs. `orElseGet` to avoid these clashes. The behavior change of `indexOf(null)` is a source incompatibility.
2. Have you considered `indexOf(predicate/object, startIndex)` like the indexOf for `String`?

Also, a significant difference between String/arrays and `List` is that `List` is structurally modifiable; the index may be no longer valid after a modification. Thus the index is not as useful. 

For example, in an [unrolled linked list](https://en.wikipedia.org/wiki/Unrolled_linked_list) implementation, each block can be independently locked for read/write access, maybe like this:
`[1, 2, 3] -> [4, 5] -> [6, 7, 8, 9]`
An index would be useless if any part of this list updates, but if the position information is stored in a `ListIterator` and we add an `beforeNext(predicate)` that moves the cursor to right before the next element that matches the predicate (or the tail if there's no match), this would be much more useful.

**Comment by Evemose** [comment]
> A few remarks:
> 
> 1. The concern on `indexOf(Object)` and `indexOf(Predicate)` is valid; `Optional` has `orElse` vs. `orElseGet` to avoid these clashes. The behavior change of `indexOf(null)` is a source incompatibility.
> 2. Have you considered `indexOf(predicate/object, startIndex)` like the indexOf for `String`?
> 
> Also, a significant difference between String/arrays and `List` is that `List` is structurally modifiable; the index may be no longer valid after a modification. Thus the index is not as useful.
> 
> For example, in an [unrolled linked list](https://en.wikipedia.org/wiki/Unrolled_linked_list) implementation, each block can be independently locked for read/write access, maybe like this: `[1, 2, 3] -> [4, 5] -> [6, 7, 8, 9]` An index would be useless if any part of this list updates, but if the position information is stored in a `ListIterator` and we add an `beforeNext(predicate)` that moves the cursor to right before the next element that matches the predicate (or the tail if there's no match), this would be much more useful.

The concern about change in behaviour of indexOf(null) is completely valid, although indexOf(null) is kind of exotic scenario. One possible workaround is to temporally (for a intermidiate period) or permanently delegate null value handing to indexOf(Object), but it seems odd to me tdb, although it preserves source compatibility.

Regarding to indexOf(predicate/object, startIndex), i thought about this, but firstly I would like to have a final and approved version of indexOf(predicate) to have a reference I`m certain in.

As for beforeNext(Predicate) for ListIterator, I havent thought about that, but now that you mentioned it, I could also consider implementing such thing. However, concern about possible index invalidation is more applicable to nature of indexOf itself. This method is primarly used when index is needed right here and now or passed to operation that is not supposed to modify list. I guess ListIterator.beforeNext and List.indexOf serve different, altough alike purpose, and therefore, in my opinion, there are place for bot in Java

**Comment by xxDark** [comment]
> One possible workaround is to temporally (for an intermidiate period) or permanently delegate null value handing to indexOf(Object), which seems odd for me to me tbh, although it preserves source compatibility.

This does not solve the problem though. Source compatibility would still be broken as one would still need to add casts everywhere, or am I misinterpreting something?
> he concern about change in behaviour of indexOf(null) is completely valid, although indexOf(null) is kind of exotic scenario.

See [this](https://github.com/search?q=%22indexOf%28null%29%22+language%3AJava+&type=code) and [that](https://github.com/search?q=%22lastIndexOf%28null%29%22+language%3AJava+&type=code) query.

**Comment by Evemose** [comment]
> This does not solve the problem though. Source compatibility would still be broken as one would still need to add casts everywhere, or am I misinterpreting something?

The code you have provided in example is still valid and non-ambigous, even without cast. The default behaviour of java compiler is to choose method where parameter at corresponding position is closest type upwards in class hierarchy. null is considered "bottom" type, which means it is at the bottom of simultaneously every class hierarchy. Compiler looks upward in class hierarchy and first class it encounters is Predicate, as Object is superclass of Predicate, and chooses indexOf(Predicate) as method to invoke. When casting null to type, compiler treats null as object of casted type rather then bottom type. However, if changes are delivered as is, default behaviour of indexOf(null) will be changed, which is not desired, of course, and, although delegation to indexOf(Object) in case of null as passed value is counterintuitive, it will preserve the current state of things.

> See [this](https://github.com/search?q=%22indexOf%28null%29%22+language%3AJava+&type=code) and [that](https://github.com/search?q=%22lastIndexOf%28null%29%22+language%3AJava+&type=code) query.

Thats a valid point, haven`t though about qurying where null is valid value.

**Comment by Evemose** [comment]
Question to reviewers: what do you think would be the best way to handle issue discussed above?

**Comment by ExE-Boss** [comment]
I’m in favour of using `findIndex`, which also avoids the issue with `List<Predicate<…>>::indexOf` being ambiguous and having binary incompatible bridge methods.

File: src/java.base/share/classes/java/util/LinkedList.java
Comment (xxDark) [inline src/java.base/share/classes/java/util/LinkedList.java:1519]: Should be `findIndex`? Similarly with `findLastIndex`.

Comment (Evemose) [reply src/java.base/share/classes/java/util/LinkedList.java]: > Should be `findIndex`? Similarly with `findLastIndex`.

Yeah i guess IDE didnt do a great job with refactor. I will review aall changes manually a bit later

**Comment by liach** [comment]
Just a side note, this patch won't be reviewed by official JDK engineers until it has rfr tag (means this PR is not draft and the automatic checks of issue, whitespace, oca etc. are passed)

You should post on `core-libs-dev@openjdk.org` to get feedbacks on the API design, too.

**Comment by minborg** [comment]
Adding (default) methods to `List` would have significant compatibility ramifications. Also, the PR needs to contain significant testing of the new proposed methods.

**Comment by Evemose** [comment]
@minborg Sorry to bother you with this kind of question, but i cant manage to find where exactly and how I can file CSR review request. Could you help me out?

**Comment by liach** [comment]
@Evemose CSR requests can only be filed by users who have a JBS account. I am glad to help you file one, but before that, you need to ensure that the API surface (the public method/fields' names, modifiers, types, and javadocs) are all finalized and won't change.

The CSR looks like this:
Summary: A concise summary about the changes, like "Add findIndex(Predicate) to List"
Problem: Description of the problem, as in why this CSR is needed
Solution: The solution to the said problem, i.e. what your task did
Specification: The git diff, but you only need to include changes to public method/field/classes' names, modifiers, types, and javadocs; you don't need to include any implementation code changes
Also, you might include things like "alternatives" describing alternative approaches and why you elect not to choose them, etc.

**Comment by AlanBateman** [comment]
> but i cant manage to find where exactly and how I can file CSR review request. Could you help me out?

I think you can ignore CSR for now, the first step when proposing an API in this area is to start a discussion on core-libs-dev to make the case and get support. The java.util.List interface dates from JDK 1.2 and there are many implementations. Even if the proposal has some merit, I think you a big part of the analysis required will be doing analysis on implementation to see what compatibility issues might come up.

**Comment by Evemose** [comment]
@liach i did revert changes where it was possible but in few files it will be just too resource-consuming to revert indentations to previous state. If this will be crucial, ill deal with it before merging

**Comment by Evemose** [comment]
I requested post in core-libs-dev. I guess I will wait a week or two to gather feedback of community before filing for CSR

**Comment by RogerRiggs** [comment]
Please revert the formatting as soon as possible; its noise as far as reviewers are concerned. 
You want reviewers to focus on the suitability of the API and clarity of the specification first, then later on the implementation.  Thanks

**Comment by Evemose** [comment]
@RogerRiggs Been done. It turned out much easier then i though it would

**Comment by xxDark** [comment]
I noticed that most (if not all) methods don't ensure non-nullability of `filter` so NPE would only be thrown if the list is not empty.

**Comment by Evemose** [comment]
> I noticed that most (if not all) methods don't ensure non-nullability of `filter` so NPE would only be thrown if the list is not empty.

Yeah, thats true. not sure if it has to throw NPE even if list is emply

**Comment by xxDark** [comment]
> > I noticed that most (if not all) methods don't ensure non-nullability of `filter` so NPE would only be thrown if the list is not empty.
> 
> Yeah, thats true. not sure if it has to throw NPE even if list is emply

Yes, it does. If it shouldn't, then why isn't code in the java.util.List is like this:
```java
default int findIndex(Predicate<? super T> filter) {
	ListIterator<T> iterator = listIterator();
	if (!iterator.hasNext()) return -1;
	Objects.requireNonNull(filter);
	int index = 0;
	do {
		if (filter.test(iterator.next()))
			return index;
		index++;
	} while (iterator.hasNext());
	return -1;
}
```
Also see methods like `removveIf`. All methods should do checks even if they essentially do nothing.

**Comment by Evemose** [comment]
> > > I noticed that most (if not all) methods don't ensure non-nullability of `filter` so NPE would only be thrown if the list is not empty.
> > 
> > 
> > Yeah, thats true. not sure if it has to throw NPE even if list is emply
> 
> Yes, it does. If it shouldn't, then why isn't code in the java.util.List is like this:
> 
> ```java
> default int findIndex(Predicate<? super T> filter) {
> 	ListIterator<T> iterator = listIterator();
> 	if (!iterator.hasNext()) return -1;
> 	Objects.requireNonNull(filter);
> 	int index = 0;
> 	do {
> 		if (filter.test(iterator.next()))
> 			return index;
> 		index++;
> 	} while (iterator.hasNext());
> 	return -1;
> }
> ```
> 
> Also see methods like `removveIf`. All methods should do checks even if they essentially do nothing.

ok, i will fix this in a while

**Comment by Evemose** [comment]
@xxDark Added null asserions everywhere except sublists that delegate execution to other lists, except for SynchroziedLsit so it doesnt acquire monitor with invalid predicate. I think this would be the best choice

File: src/java.base/share/classes/java/util/LinkedList.java
Comment (liach) [inline src/java.base/share/classes/java/util/LinkedList.java:1507]: Hmm, this view is supposed to be reversed... so last and first need a swap. This is a bug in the current codebase, a good bug to fix indeed.

Comment (liach) [reply src/java.base/share/classes/java/util/LinkedList.java:1507]: Ignore my fault, it's just delegating methods to 2 different, already reversed views...

**Comment by Evemose** [comment]
@liach Seems like discussion in mails has came to some conclusion (or people just arent interested in it anymore). Therefore, i think its time to file for csr. Here is a draft request text, but Im not sure how to fill in some fields:
Summary: Add findIndex(Predicate) and findLastIndex(Predicate) to List
Problem: Modification of java.utilList interface by adding findIndex(Predicate) and findLastIndex(Predicate) methods might introduce behavioural or byteecode incompatibilities
Solution: // dont know what to write here tbh
Specification: // do i have to paste changed code here, or provide links, or what?
Alternatives: syntax line indexOf(Predicate) and lastIndexOf(Predicate) was concidered, but discarded due to behavioural incompatibilities with invokation indexOf(null)

**Review by ExE-Boss — COMMENTED** [review]
These can be `private` as they’re not used outside the `ArrayList` class nest:

File: src/java.base/share/classes/java/util/ArrayList.java
Comment (Evemose) [reply src/java.base/share/classes/java/util/ArrayList.java:380]: Yeah i though about it but indexOfRange arent private here so either there is a point in it or its just legacy without any particular meaning

Comment (liach) [reply src/java.base/share/classes/java/util/ArrayList.java:380]: It is legacy to avoid bridge generation from the SubList.

Before introduction of nestmates in JDK 11, private methods and fields called by inner classes should be declared package-private, as javac has to generate bridge methods at each call site to abide to JVM rules (inner classes are just another class in the package so couldn't call private methods).

**Comment by Evemose** [comment]
Commited proposed changes. Also still would appreciate any help about what to write it csr template (you can see what I figured a few messages above)

**Comment by Evemose** [comment]
Seems like run failed due to internal problems of openjdk. I will trigger rerun a bit later.

**Pull request closed without merging**
